### PR TITLE
Maria Pogosjan - project-happy-thoughts-api

### PR DESCRIPTION
**Pull request by MariaPogosjan**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,21 @@
...
-Start by briefly describing the assignment in a sentence or two. Keep it short and to the point.
+## Endpoints
+- GET '/': returns the list of all endpoints;
+- GET '/thoughts': returns a list of all;
+- POST '/thoughts': posts for message, username and hashtag;
+- POST '/'/thoughts/:thoughtId/like': posts for likes; 
+
+## View it live
+- BE: https://poggi-happy-thoughts-api.herokuapp.com/
+- FE: https://poggi-happy-thoughts.netlify.app/
```
Comment (klimenko3007) [inline README.md:17]: Great documentation! Very clear description of en points. Only the last one I think has too many slashes.

File: server.js [hunk 2]
```diff
@@ -1,29 +1,131 @@
...
 const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/happyThoughts"
-mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
+// eslint-disable-next-line max-len
+mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true, useFindAndModify: false })
 mongoose.Promise = Promise
 
-// Defines the port the app will run on. Defaults to 8080, but can be 
-// overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
```
Comment (klimenko3007) [inline server.js:13]: Great cleaning up of the unnecessary comments!

File: server.js [hunk 3]
```diff
@@ -1,29 +1,131 @@
...
+const thoughtSchema = new mongoose.Schema({
+  message: {
+    type: String, 
+    required: [true, 'Message is required!'], 
+    trim: true,
+    minlength: [5, 'You have to have at least 5 character and I just got {VALUE}'],
```
Comment (klimenko3007) [inline server.js:16]: Cool, I did not know you can add a message like this to validations

File: server.js [hunk 4]
```diff
@@ -1,29 +1,131 @@
...
+    size = 20
+  }
+  
+  try {
+    const thoughts = await Thought
+      .find()
+      .limit(Number(size))
+      .skip((page - 1) * size)
+      .sort({ createdAt: 'desc' })
+      .exec()
```
Comment (klimenko3007) [inline server.js:65]: Really clean code. Great to have it like this rather than in one line!

File: server.js [hunk 5]
```diff
@@ -1,29 +1,131 @@
...
+
+app.get('/thoughts', async (req, res) => {
+  let { page, size } = req.query
+  
+  if (!page) {
+    page = 1
+  }
+  if (!size) {
+    size = 20
+  }
```
Comment (klimenko3007) [inline server.js:57]: Really cool set up for default values! I am learning a lot doing this review :D

File: server.js [hunk 6]
```diff
@@ -1,29 +1,131 @@
...
+  const tags = req.body.message.trim().split(' ').map((item) => ` #${item}`)
+
+  const tagsUnique = []
+
+  tags.forEach((item) => {
+    if (!tagsUnique.includes(item)) {
+      tagsUnique.push(item)
+    } 
+  }) 
+
```
Comment (klimenko3007) [inline server.js:87]: Very interesting way of defining tags :)

**Review by klimenko3007 — APPROVED** [review]
Great job, Maria!! Really impressive add on to the blue level requirements! Congratulations!

File: README.md [hunk 7]
```diff
@@ -1,13 +1,21 @@
...
+- How to build a registration flow
+- How to handle authentication, both in the frontend and in the backend
+- How to build a frontend and backend at the same time
 
-Start by briefly describing the assignment in a sentence or two. Keep it short and to the point.
+## Endpoints
+- GET '/': returns the list of all endpoints;
+- GET '/thoughts': returns a list of all;
+- POST '/thoughts': posts for message, username and hashtag;
+- POST '/'/thoughts/:thoughtId/like': posts for likes;
```
Comment (KarolCislowski) [inline README.md:13]: Nice documentation 👍🏻

File: server.js [hunk 8]
```diff
@@ -1,29 +1,126 @@
...
-// overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
+const thoughtSchema = new mongoose.Schema({
+  message: {
+    type: String, 
+    required: [true, 'Message is required!'], 
+    trim: true,
+    minlength: [5, 'You have to have at least 5 character and I just got {VALUE}'],
+    maxlength: 140
```
Comment (KarolCislowski) [inline server.js:17]: Nicely configured field 👍🏻

File: server.js [hunk 9]
```diff
@@ -1,29 +1,126 @@
...
+    
+  const countAllThoughts = await Thought.countDocuments() 
+
+  try {
+    const thoughts = await Thought
+      .find()
+      .limit(Number(size))
+      .skip((page - 1) * size)
+      .sort({ createdAt: 'desc' })
+      .exec()
```
Comment (KarolCislowski) [inline server.js:60]: 👍🏻

File: server.js [hunk 10]
```diff
@@ -1,29 +1,126 @@
...
+app.post('/thoughts', async (req, res) => {
+  const tags = req.body.message.trim().split(' ').map((item) => ` #${item}`)
+
+  const tagsUnique = []
+
+  tags.forEach((item) => {
+    if (!tagsUnique.includes(item)) {
+      tagsUnique.push(item)
+    } 
+  })
```
Comment (KarolCislowski) [inline server.js:81]: Maybe it would be nice to check if there are hashtags at all first?

File: server.js [hunk 11]
```diff
@@ -1,29 +1,126 @@
...
+        _id: thoughtId
+      }, 
+      { 
+        $inc: { 
+          hearts: 1 
+        } 
+      }, 
+      {
+        new: true
+      }
```
Comment (KarolCislowski) [inline server.js:112]: 💯

**Review by KarolCislowski — APPROVED** [review]
Great API 👍🏻

**Pull request closed without merging**